test/utils/structs/Vector.test.js

```diff
@@ -0,0 +1,91 @@
+const { expectEvent, expectRevert } = require('@openzeppelin/test-helpers');
+const { expect } = require('chai');
+
+const Bytes32VectorMock = artifacts.require('Bytes32VectorMock');
+
+async function getContent (vector) {
+  const length = await vector.length().then(bn => bn.toNumber());
+  const values = await Promise.all(Array(length).fill().map((_, i) => vector.at(i)));
+  return values;
+}
+
+contract('Vector', function (accounts) {
+  // Bytes32Vector
+  describe('EnumerableVector', function () {
+    const bytesA = '0xdeadbeef'.padEnd(66, '0');
+    const bytesB = '0x0123456789'.padEnd(66, '0');
+    const bytesC = '0x42424242'.padEnd(66, '0');
+    const bytesD = '0x171717'.padEnd(66, '0');
+
+    beforeEach(async function () {
+      this.vector = await Bytes32VectorMock.new();
+    });
+
+    it('empty vector', async function () {
+      expect(await this.vector.empty()).to.be.equal(true);
+      expect(await getContent(this.vector)).to.have.all.members([]);
+    });
+
+    describe('with content', function () {
+      beforeEach(async function () {
+        await this.vector.pushBack(bytesB);
+        await this.vector.pushFront(bytesA);
+        await this.vector.pushBack(bytesC);
+      });
+
+      it('content is valid', async function () {
+        expect(await this.vector.empty()).to.be.equal(false);
+        expect(await this.vector.front()).to.be.equal(bytesA);
+        expect(await this.vector.back()).to.be.equal(bytesC);
+        expect(await getContent(this.vector)).to.have.all.members([ bytesA, bytesB, bytesC ]);
+      });
+
+      it('out of bound access', async function () {
+        await expectRevert.unspecified(this.vector.at(10));
```

Comment:
It's bad that we don't have a way of testing custom errors... What if we introduce Ethers.js in this file? I think it allows testing custom errors, right?

Reply:
Would you be using pure ethers for this file, or have a mix of ethers and truffle+web3js. Using ethers + waffle would be clearer, but it would add dependencies, and would make this file look very different from the other ones.

Maybe we can do a first PR with just the vector thing (and no custom error testing) and then do a second PR with the test migration to ethers (so we don't discuss two mostly orthogonal things in the same PR).